shtdlib.sh

```diff
@@ -126,6 +126,36 @@ function in_array {
     return 1
 }
 
+
+# Returns the number of arguments passed to it
+function count_arguments {
+    echo ${#:-0}
+}
+
+# Prints the number of elements in an array using the name passed as an
+# argument in a bash version agnostic way.
+# This is important because of changes in handling of empty arrays with the -u
+# flag set which was different from bash v 4.0 until 4.4
+function count_array_elements {
+    shopt_decorator_option_name='nounset'
+    shopt_decorator_option_value='false'
+    # shellcheck disable=2015
+    shopt_decorator "${FUNCNAME[0]}" "${@:-}" && return || conditional_exit_on_fail 121 "Failed to run ${FUNCNAME[0]} with shopt_decorator"
+
+    array_ref="${1}[*]"
+    #shellcheck disable=2086
+    count_arguments ${!array_ref}
```

Comment:
`s/${!array_ref}/"${!array_ref[@]}"/` (explodes to each element quoted individually) to remove the shellcheck error

Reply:
No because they'll be individual strings surrounded by `"` (e.g. `"one" "two" "three"`). `"arr[*]"` - explodes to `"one two three"`

Reply:
```
Similar to mass-expanding positional parameters, this expands to all elements. If unquoted, both subscripts * and @ expand to the same result, if quoted, @ expands to all elements individually quoted, * expands to all elements quoted as a whole.
```

Reply:
Have you tried it?

The reason why we're counting $@ is because $@ has special handling when the "array" is empty so it does not error out while accessing it. Any other array gives an error if it's empty since that's considered undefined in some bash versions.

Reply:
Also because we're using indirection things are a bit different, with other words putting quotes around that breaks things ....